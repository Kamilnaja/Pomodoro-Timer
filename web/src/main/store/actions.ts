export const SETTINGS_SAVED = "SETTINGS_SAVED";
