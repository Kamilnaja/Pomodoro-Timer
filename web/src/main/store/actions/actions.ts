export enum MainActions {
  SAVE_POMODORO = "SAVE_POMODORO",
  SAVE_POMODORO_SUCCESS = "SAVE_POMODORO_SUCCESS",
  SAVE_POMODORO_ERROR = "SAVE_POMODORO_ERROR",
}
