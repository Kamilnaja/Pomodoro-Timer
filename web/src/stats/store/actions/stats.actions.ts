import { Action } from "redux";
import { config } from "shared/settings/initialConfig";
import { ActionWithPayload } from "shared/store/interfaces/actions/action.interface";
import { store } from "shared/store/reducers/reducer";
import { AuthError } from "../../../../../types/interfaces";
import StatsSearchResult from "../../../../../types/statistics.interfaces";

export enum StatsAction {
  GET_LAST_STATISTICS = "GET_LAST_STATISTICS",
  GET_LAST_STATISTICS_SUCCESS = "GET_LAST_STATISTICS_SUCCESS",
  GET_LAST_STATISTICS_ERROR = "GET_LAST_STATISTICS_ERROR",
}

export const getLastStatistics = (payload: number): ActionWithPayload<StatsAction, number> => ({
  type: StatsAction.GET_LAST_STATISTICS,
  payload,
});

export const getLastStatisticsSuccess = (payload: StatsSearchResult): ActionWithPayload<StatsAction, StatsSearchResult> => ({
  type: StatsAction.GET_LAST_STATISTICS_SUCCESS,
  payload,
});

export const getLastStatisticsError = (error: AuthError): ActionWithPayload<StatsAction, AuthError> => ({
  type: StatsAction.GET_LAST_STATISTICS_ERROR,
  payload: error,
});
// thunk

export const getLastStats = (duration: number) => async (dispatch: (args: Action) => void) => {
  dispatch(getLastStatistics(duration));

  makeGetStatsRequest()
    .then((payload: StatsSearchResult) => {
      dispatch(getLastStatisticsSuccess(payload));
    })
    .catch(err => {
      dispatch(getLastStatisticsError(err));
    });
};

export const makeGetStatsRequest = async () => {
  const token = store.getState().auth.token;

  const response = await fetch(`${config.url.API_URL}/stats`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  const responseBody = await response.json();

  return response.ok ? Promise.resolve(responseBody) : Promise.reject(responseBody);
};
