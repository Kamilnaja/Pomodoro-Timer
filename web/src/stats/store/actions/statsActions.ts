import { Action } from 'redux';
import { StatsSearchResult, PomodorosSearchResult } from '../../../../../types/statsInterfaces';
import { fetchData, updateData } from '../../../shared/scripts/requests';
import { getCurrentDay, getCurrentMonth, getCurrentYear } from '../../../shared/scripts/utils';
import {
  GET_STATS,
  GET_STATS_ERROR,
  GET_STATS_SUCCESS,
  GET_STATS_WITH_TAGS,
  GET_STATS_WITH_TAGS_ERROR,
  GET_STATS_WITH_TAGS_SUCCESS,
  GET_TODAY_STATS,
  GET_TODAY_STATS_ERROR,
  GET_TODAY_STATS_SUCCESS,
  SAVE_POMODORO,
  SAVE_POMODORO_ERROR,
  StatsActionsTypes,
} from './statsActionsTypes';

export const savePomodoro = (): StatsActionsTypes => ({
  type: SAVE_POMODORO,
});

const savePomodoroError = (error: any): StatsActionsTypes => ({
  type: SAVE_POMODORO_ERROR,
  payload: error,
});

const getStats = (): Action => ({
  type: GET_STATS,
});

const getStatsSuccess = (payload: StatsSearchResult): StatsActionsTypes => ({
  type: GET_STATS_SUCCESS,
  payload,
});

const getStatsError = (payload: StatsSearchResult): StatsActionsTypes => ({
  type: GET_STATS_ERROR,
  payload,
});

const getTodayStats = (): StatsActionsTypes => ({
  type: GET_TODAY_STATS,
});

const getTodayStatsSuccess = (payload: number): StatsActionsTypes => ({
  type: GET_TODAY_STATS_SUCCESS,
  payload,
});

const getTodayStatsError = (payload: any): StatsActionsTypes => ({
  type: GET_TODAY_STATS_ERROR,
  payload,
});

const getStatsWithTags = () => ({
  type: GET_STATS_WITH_TAGS,
});

const getStatsWithTagsSuccess = (payload: PomodorosSearchResult) => ({
  type: GET_STATS_WITH_TAGS_SUCCESS,
  payload,
});

export const getStatsWithTagsError = (payload: any) => ({
  type: GET_STATS_WITH_TAGS_ERROR,
  payload,
});
// thunk

export const handleGetStatsWithTags = (year: number, month: number) => async (dispatch: (args: Action) => void) => {
  dispatch(getStatsWithTags());

  fetchData(`stats/tags/${year}/${month}`)
    .then((payload: PomodorosSearchResult) => dispatch(getStatsWithTagsSuccess(payload)))
    .catch(err => dispatch(getStatsWithTagsError(err)));
};

export const handleGetStats = (year: number, month: number) => async (dispatch: (args: Action) => void) => {
  dispatch(getStats());

  fetchData(`stats/${year}/${month}`)
    .then((payload: StatsSearchResult) => dispatch(getStatsSuccess(payload)))
    .catch(err => dispatch(getStatsError(err)));
};

export const handleGetTodayStats = () => async (dispatch: (args: Action) => void) => {
  dispatch(getTodayStats());

  fetchData(`stats/${getCurrentYear()}/${getCurrentMonth()}/${getCurrentDay()}`)
    .then((payload: StatsSearchResult) =>
      dispatch(getTodayStatsSuccess(payload.pomodoros[0] ? payload.pomodoros[0].count : 0)),
    )
    .catch(err => dispatch(getTodayStatsError(err)));
};

export const handleSavePomodoro = () => async (dispatch: (arg: Action | any) => void) => {
  dispatch(savePomodoro());

  updateData('stats', {}, 'POST')
    .then(() => dispatch(handleGetTodayStats()))
    .catch(err => dispatch(savePomodoroError(err)));
};
