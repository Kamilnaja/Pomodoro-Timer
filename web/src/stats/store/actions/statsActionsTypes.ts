import { StatsSearchResult } from '../../../../../types/statisticsInterfaces';

export const GET_STATISTIC_IN_PERIOD = 'GET_STATISTIC_IN_PERIOD';
export const GET_STATISTIC_IN_PERIOD_SUCCESS = 'GET_STATISTIC_IN_PERIOD_SUCCESS';
export const GET_STATISTIC_IN_PERIOD_ERROR = 'GET_STATISTIC_IN_PERIOD_ERROR';

export const GET_TODAY_STATS = 'GET_TODAY_STATS';
export const GET_TODAY_STATS_SUCCESS = 'GET_TODAY_STATS_SUCCESS';
export const GET_TODAY_STATS_ERROR = 'GET_TODAY_STATS_ERROR';

export const SAVE_POMODORO = 'SAVE_POMODORO';
export const SAVE_POMODORO_SUCCESS = 'SAVE_POMODORO_SUCCESS';
export const SAVE_POMODORO_ERROR = 'SAVE_POMODORO_ERROR';

export interface GetStatisticsInPeriodAction {
  type: typeof GET_STATISTIC_IN_PERIOD;
}

export interface GetStatisticsInPeriodSuccessAction {
  type: typeof GET_STATISTIC_IN_PERIOD_SUCCESS;
  payload: StatsSearchResult;
}

export interface GetStatisticsInPeriodErrorAction {
  type: typeof GET_STATISTIC_IN_PERIOD_ERROR;
  payload: any;
}

export interface GetTodayStatsAction {
  type: typeof GET_TODAY_STATS;
}

export interface GetTodayStatsSuccessAction {
  type: typeof GET_TODAY_STATS_SUCCESS;
  payload: number;
}

export interface GetTodayStatsErrorAction {
  type: typeof GET_TODAY_STATS_ERROR;
  payload: any;
}

export interface SavePomodoroAction {
  type: typeof SAVE_POMODORO;
}

export interface SavePomodoroSuccessAction {
  type: typeof SAVE_POMODORO_SUCCESS;
}

export interface SavePomodoroErrorAction {
  type: typeof SAVE_POMODORO_ERROR;
  payload: any;
}

export type StatsActionsTypes =
  | GetStatisticsInPeriodAction
  | GetStatisticsInPeriodSuccessAction
  | GetStatisticsInPeriodErrorAction
  | GetTodayStatsAction
  | GetTodayStatsSuccessAction
  | GetTodayStatsErrorAction
  | SavePomodoroAction
  | SavePomodoroSuccessAction
  | SavePomodoroErrorAction;
