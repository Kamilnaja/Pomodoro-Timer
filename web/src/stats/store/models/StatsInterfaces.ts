import { PomodorosDoneInDay } from '../../../../../types/statisticsInterfaces';

export interface StatsState {
  isLoading: boolean;
  error: string;
  results: PomodorosDoneInDay[];
  todayResults: number;
}

export type HandleGetStats = (year?: number, month?: number) => void;
