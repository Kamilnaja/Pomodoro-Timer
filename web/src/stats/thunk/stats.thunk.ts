import { handleErrors } from "../../main/store/thunk/main.thunk";
import { initialConfig } from "../../shared/settings/initialConfig";
import TodayStatistics from "../../shared/store/interfaces/todayStatistics.interface";
import {
  getTodayStatistics,
  getTodayStatisticsError,
  getTodayStatisticsSuccess,
} from "../store/actions/stats.actions";

const API_URL = initialConfig.apiUrl;

const makeGetRequest = () => fetch(API_URL + "/pomodoros_done_today");

export const getPomodorosThunk = () => (dispatch: Function) => {
  dispatch(getTodayStatistics());

  return makeGetRequest()
    .then(handleErrors)
    .then((response) => response.json())
    .then((payload: TodayStatistics) => {
      return dispatch(getTodayStatisticsSuccess(payload));
    })
    .catch((error) => dispatch(getTodayStatisticsError(error)));
};
