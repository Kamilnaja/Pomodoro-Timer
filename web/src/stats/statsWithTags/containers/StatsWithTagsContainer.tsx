import React, { Component } from 'react';
import { connect } from 'react-redux';
import { StatsState } from '../../store/models/StatsInterfaces';
import { StatsWithTagsComponent } from '../components/StatsWithTagsComponent';
import { StatsWithTagsContainerProps } from './StatsWithTagsContainerProps';

class StatsWithTagsContainer extends Component<StatsWithTagsContainerProps> {
  render() {
    return <StatsWithTagsComponent />;
  }
}

const mapStateToProps = (state: { stats: StatsState }) => {
  const stats = state.stats;
  return { stats };
};

const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(StatsWithTagsContainer);
