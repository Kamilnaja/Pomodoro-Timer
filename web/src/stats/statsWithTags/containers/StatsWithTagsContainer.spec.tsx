import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { StatsActionsTypes } from '../../store/actions/statsActionsTypes';
import { StatsState } from '../../store/models/StatsInterfaces';
import { createPomodoroSearchResult, createStatsState } from '../../testing/stats.test.data';
import { ConnectedStatsWithTagsContainer } from './StatsWithTagsContainer';

const mockStore = configureStore([thunk]);

describe('StatsContainer', () => {
  let component: renderer.ReactTestRenderer;
  let store: Store<any, StatsActionsTypes>;
  const state: StatsState = createStatsState();

  beforeEach(() => {
    store = mockStore({
      stats: {
        ...state,
        isLoadingWithStats: false,
        statsWithTags: createPomodoroSearchResult(),
      },
    });

    component = renderer.create(
      <Provider store={store}>
        <ConnectedStatsWithTagsContainer />
      </Provider>,
    );
  });

  it('should match snapshot', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
