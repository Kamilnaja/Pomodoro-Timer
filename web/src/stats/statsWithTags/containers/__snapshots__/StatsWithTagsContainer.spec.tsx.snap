// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsContainer should match snapshot 1`] = `
<div
  className="table"
>
  <table
    className="stats__table table table table-sm table-striped table-bordered table-hover"
  >
    <thead
      className="table__head"
    >
      <tr>
        <th>
          Day
        </th>
        <th>
          Tags
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          2021-03-01T05:47:01.476Z
        </td>
        <td />
      </tr>
      <tr>
        <td>
          2021-03-01T06:13:36.785Z
        </td>
        <td />
      </tr>
      <tr>
        <td>
          2021-03-01T17:28:16.182Z
        </td>
        <td />
      </tr>
      <tr>
        <td>
          2021-03-02T19:59:00.403Z
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
